### PR TITLE
feat: added collapsible sidebar

### PR DESCRIPTION
**Pull request by tfu2003**
Created a collapsible sidebar. The open sidebar shows the icons and names and the closed only shows the icons. Currently, there is no delay for the transition so it may look choppy.

### CODE REVIEW

File: src/components/layout.tsx [hunk 1]
```diff
@@ -19,64 +20,123 @@ import { useRouter } from 'next/router';
 
 function Sidebar() {
     const router = useRouter();
+    const [open, setOpen] = useState(true);
 
     return (
-        <Card className="h-full w-full max-w-[20rem] p-4 shadow-blue-gray-900/5">
+        <Card className={`transition-all ${open ? "max-w-[20rem]" : "max-w-[6rem]"} p-4 shadow-blue-gray-900/5 transition-all duration-100 ease-in-out`}>
```
Comment (hanmangokiwi) [inline src/components/layout.tsx:26]: could we add the h-full back?

**Review by hanmangokiwi — CHANGES_REQUESTED** [review]

File: src/components/layout.tsx (hunk 1 above)
Comment (tfu2003) [reply src/components/layout.tsx:26]: Added h-full back.

File: src/components/layout.tsx [hunk 2]
```diff
@@ -19,64 +20,123 @@ import { useRouter } from 'next/router';
...
-                    Settings
-                </ListItem>
-                <ListItem onClick={() => void router.push('/account')}>
-                    <ListItemPrefix>
-                        <UserCircleIcon className="h-5 w-5" />
-                    </ListItemPrefix>
-                    Account
-                </ListItem>
+            <List className={`select-none ${open ? "min-w-[15rem]" : "min-w-[3rem]"} transition-all duration-100 ease-in-out`}>
+                <hr className={`my-2 border-blue-gray-50 ${open ? "max-w-[20rem]" : "max-w-[3rem]"}`}/>
```
[the remark this answers is not in the document]
Comment (tfu2003) [reply src/components/layout.tsx:40]: Made minimized sidebar slimmer. Something to note is that the icon may not be centered in the shadow when you hover over it. I fixed it by adjusting the widths, but I'm looking for a more permanent fix with centering items.

File: src/components/layout.tsx [hunk 3]
```diff
@@ -17,66 +18,108 @@ import {
...
+    const open = useContext(StateContext);
+
+    return (
+        <>
+            {open ? (
+                <ListItem
+                    onClick={() => void router.push(`/${title.toLowerCase()}`)}
+                >
+                    <ListItemPrefix>{icon}</ListItemPrefix>
+                    {title}
```
Comment (hanmangokiwi) [inline src/components/layout.tsx:41]: can we do something like {open? title : null} instead?

File: src/components/layout.tsx [hunk 4]
```diff
@@ -17,66 +18,108 @@ import {
...
+    state: boolean;
+}) {
+    const router = useRouter();
+    const open = useContext(StateContext);
+
+    return (
+        <>
+            {open ? (
+                <ListItem
+                    onClick={() => void router.push(`/${title.toLowerCase()}`)}
```
Comment (hanmangokiwi) [inline src/components/layout.tsx:38]: could we have a new prop instead of making the title lowercase?

File: src/components/layout.tsx [hunk 5]
```diff
@@ -17,66 +18,108 @@ import {
...
 
+const StateContext = createContext(true);
+
+function Content({
+    icon,
+    title,
+}: {
+    icon: React.ReactNode;
+    title: string;
+    state: boolean;
```
Comment (hanmangokiwi) [inline src/components/layout.tsx:29]: remove unused "state"

File: src/components/layout.tsx [hunk 6]
```diff
@@ -17,66 +18,108 @@ import {
...
+                <hr
+                    className={`my-2 border-blue-gray-50 ${
+                        open ? 'max-w-[20rem]' : 'max-w-[3rem]'
+                    }`}
+                />
+                <StateContext.Provider value={open}>
+                    <Content
+                        icon={<PencilIcon className="h-5 w-5" />}
+                        title={'Track'}
+                        state={open}
```
Comment (hanmangokiwi) [inline src/components/layout.tsx:90]: we already are using the StateContext provider, we don't need to pass in open to each component

**Review by hanmangokiwi — CHANGES_REQUESTED** [review]

File: src/components/layout.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (tfu2003) [reply src/components/layout.tsx]: Looks like this now.
image

File: src/components/layout.tsx (hunk 6 above)
Comment (tfu2003) [reply src/components/layout.tsx:90]: Removed.

File: src/components/layout.tsx (hunk 5 above)
Comment (tfu2003) [reply src/components/layout.tsx:29]: Removed.

File: src/components/layout.tsx (hunk 4 above)
Comment (tfu2003) [reply src/components/layout.tsx:38]: The title has to be lowercase to be able to redirect them to the appropriate page (localhost:3000/Track is different from localhost:3000/track). The actual title shown isn't changed on the sidebar. Even if it was a new prop, it'd still be the same code.

File: src/components/layout.tsx (hunk 3 above)
Comment (tfu2003) [reply src/components/layout.tsx:41]: Only thing I can think of is making a new component, which I did.

File: src/components/layout.tsx [hunk 7]
```diff
@@ -19,64 +20,123 @@ import { useRouter } from 'next/router';
...
-                    className="h-8 w-8"
-                />
-                <Typography variant="h5" color="blue-gray">
+                <button onClick={() => setOpen(!open)}>
+                    <Bars3Icon className="h-7 w-7" />
+                </button>
+                <Typography
+                    variant="h5"
+                    color="blue-gray"
+                    className={open ? '' : 'hidden'}
```
[the remark this answers is not in the document]
Comment (tfu2003) [reply src/components/layout.tsx:71]: Done. Although personally I don't think it looks more readable.

**Review by hanmangokiwi — APPROVED** [review]
LGTM!

**Pull request merged**